doc/site/content/docs/guides/rendering/making-shaders.md

```diff
@@ -0,0 +1,366 @@
+
+	vec4 fogColor; //fog color
+	vec4 fogParams; //fog {start, end, 0.0, scale}
+
+	vec4 sunDir; // (sky != nullptr) ? sky->GetLight()->GetLightDir() : float4(/*map default*/ 0.0f, 0.447214f, 0.894427f, 1.0f);
+
+	vec4 sunAmbientModel;
+	vec4 sunAmbientMap;
+	vec4 sunDiffuseModel;
+	vec4 sunDiffuseMap;
+	vec4 sunSpecularModel; // float4{ sunLighting->modelSpecularColor.xyz, sunLighting->specularExponent };
+	vec4 sunSpecularMap; //  float4{ sunLighting->groundSpecularColor.xyz, sunLighting->specularExponent };
+
+	vec4 shadowDensity; //  float4{ sunLighting->groundShadowDensity, sunLighting->modelShadowDensity, 0.0, 0.0 };
+
+	vec4 windInfo; // windx, windy, windz, windStrength
+	vec2 mouseScreenPos; //x, y. Screen space.
+	uint mouseStatus; // bits 0th to 32th: LMB, MMB, RMB, offscreen, mmbScroll, locked
+	uint mouseUnused;
+	vec4 mouseWorldPos; //x,y,z; w=0 -- offmap. Ignores water, doesn't ignore units/features under the mouse cursor
+
+	vec4 teamColor[255]; //all team colors
+};
+```
+
+### Custom Uniforms
+
+Often times, you want to pass in your own uniform data. To do that, we have to define uniform bindings when creating our shader. Our fragment shader has an input texture, so let's put that in:
+
+```lua
+shader = gl.CreateShader({
+    fragment = VFS.LoadFile("path/to/shader.frag.glsl"),
+    vertex = VFS.LoadFile("path/to/shader.vert.glsl"),
+    uniformInt = {
+        inputTexture = 0, -- we'll get to this
+    }
+})
+```
+
+The `ShaderParams` table has a number of uniform input fields; these are maps mapping the uniform name as it appears in the shader to a table of the data.
+
+#### Pass In Vec4
+
+For vectors, the data is simply an array.
+
+```lua
+uniformFloat = {
+    myVec4Uniform = {1.0, 1.0, 1.0, 1.0},
+}
+```
+
+#### Pass In Textures
+
+This one is a little trickier, but it's easy once you figure it out. Your sampler is actually passed in as an integer index:
+
+```lua
+local MY_TEXTURE_INDEX = 0
+uniformInt = {
+        myTexture = MY_TEXTURE_INDEX,
+    }
+```
+
+Then, we have to actually upload the texture data.
+
+```lua
+local texture_sampling_mode = '' -- you can set different sampling modes but I am not sure where the documentation is for this
+local filepath = texture_sampling_mode .. "path/to/my_texture.png"
+gl.Texture(filepath, MY_TEXTURE_INDEX)
+```
+
+And you should be good to go.
+
+There's some built-in path values and formats:
+
+| Name | Role |
+| --- | --- |
+| $heightmap | ground heightmap |
+| $reflection | reflection cube map |
+| $specular | specular cube map |
+| $shadow | shadowmap | 
+| !x | Lua generated texture x |
+| %x:0 | unitDef x s3o tex1 |
+| %-x:1 | featureDef x s3o tex2 |
+| #x | unitDef x build pic  |
+| ^x | unitDef x icon |
+
+### Editing Uniforms During Runtime
+
+To change uniforms during runtime, you will have to grab a handle to the uniform. Luckily, it's easy to do that.
+
+```lua
+local uniform_handle = gl.GetUniformLocation(shader, "myUniformName")
```

Comment:
This is not a handle, but a location. It's also optional as `gl.Uniform` and friends can work with uniform names too (although a bit slower).